src/rfc3161_client/verify.py

```diff
@@ -12,170 +13,267 @@
+        self._policy_id: cryptography.x509.ObjectIdentifier | None = policy_id
+        self._tsa_certificate: cryptography.x509.Certificate | None = tsa_certificate
+        self._intermediates: list[cryptography.x509.Certificate] = intermediates or []
+        self._roots: list[cryptography.x509.Certificate] = roots or []
+        self._nonce: int | None = nonce
+        self._common_name: str | None = common_name
+
+    def policy_id(self, policy_oid: cryptography.x509.ObjectIdentifier) -> VerifierBuilder:
+        """Set the policy ID."""
+        if self._policy_id is not None:
+            msg = "The policy id can be set only once"
+            raise ValueError(msg)
+        builder = copy(self)
+        builder._policy_id = policy_oid
+        return builder
+
+    def tsa_certificate(self, certificate: cryptography.x509.Certificate) -> VerifierBuilder:
+        """Set the TSA certificate."""
+        if self._tsa_certificate is not None:
+            msg = "The TSA certificate can be set only once"
+            raise ValueError(msg)
+        builder = copy(self)
+        builder._tsa_certificate = certificate
+        return builder
+
+    def add_intermediate_certificate(
+        self, certificate: cryptography.x509.Certificate
+    ) -> VerifierBuilder:
+        """Add an intermediate certificate."""
+        intermediates = self._intermediates
+        if certificate in intermediates:
+            msg = "The certificate is already present"
+            raise ValueError(msg)
+        intermediates.append(certificate)
+
+        builder = copy(self)
+        builder._intermediates = intermediates
+        return builder
+
+    def add_root_certificate(self, certificate: cryptography.x509.Certificate) -> VerifierBuilder:
+        """Add a root certificate."""
+        roots = self._roots
+        if certificate in roots:
+            msg = "The certificate is already present"
+            raise ValueError(msg)
+        roots.append(certificate)
+
+        builder = copy(self)
+        builder._roots = roots
+        return builder
+
+    def nonce(self, nonce: int) -> VerifierBuilder:
+        """Set the nonce."""
+        if nonce < 0:
+            msg = "The nonce must not be negative"
+            raise ValueError(msg)
+        if self._nonce is not None:
+            msg = "The nonce can be set only once"
+            raise ValueError(msg)
+        builder = copy(self)
+        builder._nonce = nonce
+        return builder
+
+    def common_name(self, name: str) -> VerifierBuilder:
+        """Set the common name."""
+        if self._common_name is not None:
+            msg = "The name can be set only once"
+            raise ValueError(msg)
+        builder = copy(self)
+        builder._common_name = name
+        return builder
+
+    def build(self) -> _Verifier:
+        """Build the Verifier."""
+        if not self._roots:
+            msg = "A Verifier must have at least one root certificate set."
```

Comment:
Nitpick: no periods or articles in exceptions 🙂 

```suggestion
            msg = "Verifier must have at least one root certificate set"
```